### PR TITLE
#167733651 Users should be able to make one-way trip request

### PR DESCRIPTION
**Pull request by danielufeli**
### What does this PR do?
It allows users to be able to make a one-way trip request

### Description of Task to be completed?
- Wrote tests to test functionality
- added request controller
- added trip request middleware
- added trip input validations

### How should this be manually tested?
- Clone the repo and cd into app root
- Setup necessary environment variables for testing environment (check .sample.env)
- Run docker-compose up --build to build docker image or run locally using npm run start:dev
- Run npm run pretest to migrate and seed database
- run npm run test to run all tests
- On postman, run localhost:3000/api/trip/request with trip request json parameters to post a trip request: 

### Any background context you want to provide?
none

### screenshots (if appropriate)
image

### What are the relevant pivotal tracker stories?
[#167733651](https://www.pivotaltracker.com/story/show/167733651)

### CODE REVIEW

**Review by Baystef — APPROVED** [review]

File: src/controllers/requestController.js [hunk 1]
```diff
@@ -31,4 +31,22 @@ export default class RequestController {
...
+   * @param {Request} req - The request from the endpoint.
+   * @param {Response} res - The response returned by the method.
+   * @returns { JSON } - A JSON object containing success or failure details.
+   * @memberof RequestController
+   */
+  static async oneWayTripRequest(req, res) {
+    try {
+      const { body } = req;
+      const oneWayTrip = await createTripRequest({ ...body });
+      return successResponse(res, oneWayTrip, 201);
```
Comment (Baystef) [inline src/controllers/requestController.js]:
```suggestion
      delete oneWayTrip.returnDate;
      return successResponse(res, oneWayTrip, 201);
```

File: src/middlewares/tripRequestMiddleware.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+     * @param {object} req - The request from the endpoint.
+     * @param {object} res - The response returned by the method.
+     * @param {object} next - Call the next operation.
+     * @returns {object} - Returns an object (error or response).
+     */
+  static async onTripRequest(req, res, next) {
+    try {
+      const validated = await tripRequest(req.body);
+      const { token } = req.cookies;
+      const { email } = verifyToken(token);
```
Comment (Segun-Ogundipe) [inline src/middlewares/tripRequestMiddleware.js]: We now have a middleware for verifying tokens, add that to the route declaration then get the user's email from the user object that's added to `req` by the middleware. You won't need to call `verifyToken here

File: src/middlewares/tripRequestMiddleware.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      if (validated) {
+        const user = await find({ email });
+        if (user) {
+          req.body.requesterId = user.id;
+          return next();
+        }
+        errorResponse(res, { code: 404, message: 'User does not exist' });
```
Comment (Segun-Ogundipe) [inline src/middlewares/tripRequestMiddleware.js]: Throw ApiError here
```suggestion
        throw new ApiError(404, 'User does not exist');
```

File: src/middlewares/tripRequestMiddleware.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      }
+    } catch (error) {
+      errorResponse(res, { code: 400, message: error.details[0].context.label });
```
Comment (Segun-Ogundipe) [inline src/middlewares/tripRequestMiddleware.js]: We should make this dynamic
```suggestion
      errorResponse(res, { code: error.status || 500, message: error.message });
```

File: src/test/request.test.js [hunk 5]
```diff
@@ -74,4 +80,62 @@ describe('Request route endpoints', () => {
...
     });
   });
+
+  describe('Trip Request Endpoint', () => {
+    it('should successfully create a one-way trip request', async () => {
+      const response = await chai
+        .request(server).post('/api/trip/request').set('Cookie', `token=${adminToken};`)
+        .send(tripRequest);
+      expect(response).to.have.status(201);
+      expect(response.body.data).to.be.a('object');
```
Comment (Segun-Ogundipe) [inline src/test/request.test.js]: We should use `include`
```suggestion
      expect(response.body.data).to.include({//some key/value pairs of the expected object});
```
Do this everywhere this method is used

File: src/validation/tripRequestValidation.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+       * @param {object} tripObject - The request object
+       * @param {object} res - The request response object
+       * @returns {object} - returns an object (error or response).
+       */
+  static async tripRequest(tripObject) {
+    const schema = {
+      tripType: joi.string().valid('One-way', 'Round-Trip', 'Multi-leg').required()
+        .label('Please select a trip type'),
+      purpose: joi.string().min(3).max(25).required()
+        .label('Please enter a valid purpose \n the field must not be empty and it must be more than 3 letters'),
```
Comment (Segun-Ogundipe) [inline src/validation/tripRequestValidation.js]: We should return specific error messages. That helps the user understands what they are getting wrong
Please check how this is done in  #`BookingValidator`

**Review by Segun-Ogundipe — CHANGES_REQUESTED** [review]
Great job!!! I left a couple of comments that should be responded to before merge

File: src/middlewares/tripRequestMiddleware.js (hunk 2 above)
Comment (danielufeli) [reply src/middlewares/tripRequestMiddleware.js]: ok changed

File: src/middlewares/tripRequestMiddleware.js [hunk 7]
```diff
@@ -20,12 +20,15 @@ export default class TripRequestMiddleware {
...
-      const { email } = verifyToken(token);
+      const { email } = req.data;
       if (validated) {
         const user = await find({ email });
         if (user) {
           req.body.requesterId = user.id;
+          req.data.firstName = user.firstName;
+          req.data.lastName = user.lastName;
+          req.data.gender = user.gender;
+          req.data.lineManager = user.lineManager;
```
Comment (Segun-Ogundipe) [inline src/middlewares/tripRequestMiddleware.js]: We could make line 27 - 31 into an object
```
const requesterObj = {
    requesterId = user.id,
    requesterFirstName = user.firstName,
    requesterLastName = user.lastName,
    ...
};

    req.requester = requesterObj;
```

File: src/middlewares/tripRequestMiddleware.js (hunk 3 above)
Comment (Segun-Ogundipe) [reply src/middlewares/tripRequestMiddleware.js]: This was not responded to. We have the custom error class for cases such as this

**Review by Segun-Ogundipe — CHANGES_REQUESTED** [review]
Left some comments

File: src/test/request.test.js (hunk 5 above)
Comment (dchima) [reply src/test/request.test.js]: Hey, @Segun-Ogundipe , I need a few clarificiations on your comment. Is the use of `include` call in place of `to.be.a()` a hard rule? I don't see the problem with checking if the data is an object, and then also checking for key pairs included in the object. `to.be.a()` shouldn't be changed outright, except theres a reason not to use that call.

File: src/middlewares/tripRequestMiddleware.js (hunk 3 above)
Comment (danielufeli) [reply src/middlewares/tripRequestMiddleware.js]: I have updated it.

File: src/middlewares/tripRequestMiddleware.js (hunk 4 above)
Comment (danielufeli) [reply src/middlewares/tripRequestMiddleware.js]: ok done thank you

File: src/validation/tripRequestValidation.js (hunk 6 above)
Comment (danielufeli) [reply src/validation/tripRequestValidation.js]: done

File: src/middlewares/tripRequestMiddleware.js (hunk 7 above)
Comment (danielufeli) [reply src/middlewares/tripRequestMiddleware.js]: done

**Review by Segun-Ogundipe — APPROVED** [review]
Nicely done

File: src/validation/tripRequestValidation.js [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    }
+    return true;
+  }
+
+  /**
+   * Validates departureDate and returnDate keys
+   * @param {string} key - The key to validate
+   * @returns {Error} Returns a descriptive error message
+   */
+  static validateAccDate(key) {
```
Comment (dchima) [inline src/validation/tripRequestValidation.js]: Can you please change this method name to not be abbreviated.

Comment (danielufeli) [reply src/validation/tripRequestValidation.js]: ok done

File: src/test/request.test.js (hunk 5 above)
Comment (amarachukwu-agbo) [reply src/test/request.test.js]: This is to enable us check the contents of the object to ensure we are getting what we expect. {} and {key: 'value'} are both objects and will pass the test you have written even though the first object does not contain what we are expecting.

File: src/controllers/requestController.js [hunk 9]
```diff
@@ -31,4 +31,24 @@ export default class RequestController {
...
+   * @param {Response} res - The response returned by the method.
+   * @returns { JSON } - A JSON object containing success or failure details.
+   * @memberof RequestController
+   */
+  static async oneWayTripRequest(req, res) {
+    try {
+      const { body } = req;
+      const { requester } = req;
+      const oneWayTrip = await createTripRequest({ ...body });
+      delete oneWayTrip.returnDate;
```
Comment (amarachukwu-agbo) [inline src/controllers/requestController.js]: Shouldn't the delete happen before the creation on line 47?

File: src/middlewares/tripRequestMiddleware.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+     * @param {object} res - The response returned by the method.
+     * @param {object} next - Call the next operation.
+     * @returns {object} - Returns an object (error or response).
+     */
+  static async tripCheckUser(req, res, next) {
+    try {
+      const {
+        requesterfirstName, requesterlastName, requestergender, requesterlineManager
+      } = req.requester;
+      if (requesterfirstName === null) {
```
Comment (amarachukwu-agbo) [inline src/middlewares/tripRequestMiddleware.js]: These checks for null can be shortened to `if !requesterName`

Comment (amarachukwu-agbo) [reply src/middlewares/tripRequestMiddleware.js]: We can also refactor this  method and make it less repititive

File: src/controllers/requestController.js (hunk 9 above)
Comment (danielufeli) [reply src/controllers/requestController.js]: resolved @amarachukwu-agbo

File: src/middlewares/tripRequestMiddleware.js (hunk 10 above)
Comment (danielufeli) [reply src/middlewares/tripRequestMiddleware.js]: done thank you

File: src/test/request.test.js (hunk 5 above)
Comment (danielufeli) [reply src/test/request.test.js]: ok thank you @amarachukwu-agbo

**Review by amarachukwu-agbo — COMMENTED** [review]
LGTM

**Pull request merged**